octopusdeploy/lifecycles.go

```diff
@@ -0,0 +1,204 @@
+package octopusdeploy
+
+import (
+	"fmt"
+	"gopkg.in/go-playground/validator.v9"
+	"net/url"
+
+	"github.com/dghubble/sling"
+)
+
+type LifecycleService struct {
+	sling *sling.Sling
+}
+
+func NewLifecycleService(sling *sling.Sling) *LifecycleService {
+	return &LifecycleService{
+		sling: sling,
+	}
+}
+
+type Lifecycles struct {
+	Items []Lifecycle `json:"Items"`
+	PagedResults
+}
+
+type Lifecycle struct {
+	ID                      string          `json:"Id,omitempty"`
+	Name                    string          `json:"Name" validate:"required"`
+	Description             string          `json:"Description,omitempty"`
+	ReleaseRetentionPolicy  RetentionPeriod `json:"ReleaseRetentionPolicy"`
+	TentacleRetentionPolicy RetentionPeriod `json:"TentacleRetentionPolicy"`
+	Phases                  []Phase         `json:"Phases"`
+}
+
+type RetentionUnit string
+
+const (
+	RetentionUnit_Days = "Days"
+	RetentionUnit_Items = "Items"
+)
+
+type RetentionPeriod struct {
+	Unit              RetentionUnit `json:"Unit"`
+	QuantityToKeep    int32         `json:"QuantityToKeep"`
+	ShouldKeepForever bool          `json:"ShouldKeepForever"`
+}
+
+type Phase struct {
+	ID                                 string          `json:"Id,omitempty"`
+	Name                               string          `json:"Name" validate:"required"`
+	MinimumEnvironmentsBeforePromotion int32           `json:"MinimumEnvironmentsBeforePromotion"`
+	IsOptionalPhase                    bool            `json:"IsOptionalPhase"`
+	ReleaseRetentionPolicy             *RetentionPeriod `json:"ReleaseRetentionPolicy"`
+	TentacleRetentionPolicy            *RetentionPeriod `json:"TentacleRetentionPolicy"`
+	AutomaticDeploymentTargets         []string        `json:"AutomaticDeploymentTargets"`
+	OptionalDeploymentTargets          []string        `json:"OptionalDeploymentTargets"`
+}
+
+
+func NewLifecycle(name string) *Lifecycle {
+	return &Lifecycle{
+		Name:   name,
+		Phases: []Phase{},
+		TentacleRetentionPolicy:RetentionPeriod{
+			Unit:RetentionUnit_Days,
+		},
+		ReleaseRetentionPolicy:RetentionPeriod{
+			Unit:RetentionUnit_Days,
+		},
+	}
+}
+
+// ValidateLifecycleValues checks the values of a Lifecycle object to see if they are suitable for
+// sending to Octopus Deploy. Used when adding or updating lifecycles.
+func ValidateLifecycleValues(Lifecycle *Lifecycle) error {
+	validate := validator.New()
+
+	err := validate.Struct(Lifecycle)
+
+	if err != nil {
+		return err
+	}
+
+	if Lifecycle.Phases != nil {
+		for _, phase := range Lifecycle.Phases {
```

Comment:
Is there a better way of doing this instead of jumping out at the first error?

Reply:
I am planning on replacing all validation with some helper methods, you can see an example here: https://github.com/MattHodge/go-octopusdeploy/blob/master/octopusdeploy/project.go#L69

If you use the `ValidateMultipleProperties` function I can solve this in there to return ALL the errors found and not just the first one.

Reply:
https://github.com/MattHodge/go-octopusdeploy/issues/36